### PR TITLE
Support for bundled Julia

### PR DESCRIPTION
**Pull request by one-more-minute**
Putting together some bits so that we can dump Julia into the Atom app and have julia-client find it.

I'm also going to try and hack around the module precompilation issues so that we can bundle precompiled packages. Overall, I'm hoping that we can get the time to download Juno, install it and produce your first plot down to under a minute.

### CODE REVIEW

**Comment by Varanas** [comment]
Three things:
- I think it'd probably be better to use something like `atom.config.resourcePath` to navigate to the `resources` folder.
- The bundled julia should also be the default for all the other platforms, no?
- Minor naming nitpick: I'd rename `bundleExe` to `bundledExe` or something like that, then it doesn't sound like the function does or mutates something.

Generally, I think it'd be nice to check the existence of the specified executable no matter whether it is the default or not when starting julia. Then we can throw a nicer error if it fails.

**Comment by one-more-minute** [comment]
* Makes sense.
* Of course – I should probably have made it clearer that this is a WIP ;) Once things are working on OS X I'll start testing on Windows.
* Also makes sense.

Definitely agree with checking the bundled executable, will put together a seperate PR with the `which` trick I mentioned in the gitter.

**Comment by one-more-minute** [comment]
I've gone ahead and implemented the existence checking here as well. Would be good to have a quick set of eyes over this if possible, otherwise I'll merge later today!

File: lib/connection/process.coffee [hunk 1]
```diff
@@ -1,8 +1,32 @@
...
+fs = require 'fs'
 
 module.exports =
+
+  bundledExe: ->
+    res = path.dirname(atom.config.resourcePath)
+    p = if process.platform is 'darwin'
+      path.join res, 'julia/bin/julia'
+    else if process.platform is 'win32'
+      path.join res, 'julia/bin/julia.exe'
```
Comment (Varanas) [inline lib/connection/process.coffee]: Do we only have OSX and Windows bundles? The linux case should be the same as OSX, no?

File: lib/connection/process.coffee [hunk 2]
```diff
@@ -1,8 +1,32 @@
...
+    if fs.existsSync p then p
+
+  packageDir: (s...) ->
+    path.join __dirname, '..', '..', s...
+
+  initialiseClient: (f) ->
+    packageDir = path.join __dirname, '..', '..'
```
Comment (Varanas) [inline lib/connection/process.coffee]: This line is unnecessary, right?

File: lib/connection/process.coffee [hunk 3]
```diff
@@ -11,28 +35,49 @@ module.exports =
...
+    which = if process.platform is 'win32' then 'where' else 'which'
+    proc = child_process.spawn which, [path]
+    proc.on 'exit', (status) ->
+      cb status == 0
+
+  jlNotFound: (path) ->
+    atom.notifications.addError "Julia could not be found.",
+      detail: """
+      We tried to launch Julia from:
+      #{path}
```
Comment (Varanas) [inline lib/connection/process.coffee]: Maybe mention that the path can be changed in the settings?

File: lib/connection/spawnInterruptibleJulia.ps1 [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
-	[string] $jlpath, 
-	[string] $jloptions
+	[Int32] $port,
+	[string] $jlpath,
+	[string] $boot
 )
 
 # start Julia
-$proc = Start-Process $jlpath "$jloptions -e `"import Atom; @sync Atom.connect($port)`"" -NoNewWindow -PassThru
+$proc = Start-Process $jlpath @($boot, $port) -NoNewWindow -PassThru
```
Comment (Varanas) [inline lib/connection/spawnInterruptibleJulia.ps1]: Is this valid?
Also, if all goes well, this file won't survive for long and be superseded be something actually working... :)

Comment (one-more-minute) [inline lib/connection/spawnInterruptibleJulia.ps1]: Yup, works for me. Looking forward to something working!

File: lib/connection/process.coffee [hunk 5]
```diff
@@ -11,28 +32,51 @@ module.exports =
...
-      text = data.toString()
-      if text then cons.c.err text
+
+    @checkExe @jlpath(), (exists) =>
+      if not exists
+        @jlNotFound @jlpath()
+        client.cancelBoot()
+        return
+
+      if atom.config.get 'julia-client.initialiseClient'
```
Comment (Varanas) [inline lib/connection/process.coffee]: Not sure I understand this setting -- is it only set in the bundles or on first start or something?

File: lib/connection/spawnInterruptibleJulia.ps1 (hunk 4 above)
Comment (Varanas) [inline lib/connection/spawnInterruptibleJulia.ps1]: Yeah, I'll try hacking something together this evening -- Windows is a nightmare when it comes to doing something like that.

File: lib/connection/process.coffee (hunk 5 above)
Comment (one-more-minute) [inline lib/connection/process.coffee]: Yes, I'll include the setting in the bundles and then it will only be active once, after the initial install.

Comment (Varanas) [inline lib/connection/process.coffee]: Makes sense.

**Comment by Varanas** [comment]
LGTM then. Great work!

**Comment by one-more-minute** [comment]
Thanks!

File: lib/connection/process.coffee [hunk 6]
```diff
@@ -11,28 +32,51 @@ module.exports =
       process.env.HOME || process.env.USERPROFILE
 
   jlpath: -> atom.config.get("julia-client.juliaPath")
-  # TODO: this is very naïve.
-  jlargs: -> atom.config.get("julia-client.juliaArguments").split ' '
+
+  checkExe: (path, cb) ->
+    which = if process.platform is 'win32' then 'where' else 'which'
```
Comment (Varanas) [inline lib/connection/process.coffee]: @one-more-minute: This doesn't work on Windows when `path` is an actual path. I think we would need to call `where /r path` instead, but that doesn't work if `path` is only `julia`...

Once again, Windows is annoying. We should probably check for success of `spawn("$path", ["-h")` or something like that.

Comment (one-more-minute) [inline lib/connection/process.coffee]: We can just add an early return if `fs.fileExists(path)`. That will be far faster than shelling out on Windows anyway.

Comment (Varanas) [inline lib/connection/process.coffee]: Sure, that'd work as well.

**Pull request merged**